sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+  spi->tx_fifo_base = tx_fifo_start;
+  spi->tx_fifo_len = config->tx_fifo_len;
+
+  // Now that we know that all function arguments are valid, we can abort
+  // all current transactions and begin reconfiguring the device.
+  dif_spi_device_result_t err;
+  err = dif_spi_device_abort(spi);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+  err = dif_spi_device_irq_reset(spi);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_CFG_REG_OFFSET, device_config);
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_RXF_ADDR_REG_OFFSET,
+                      rx_fifo_bounds);
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_TXF_ADDR_REG_OFFSET,
+                      tx_fifo_bounds);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  // Set the `abort` bit, and then spin until `abort_done` is asserted.
+  mmio_region_nonatomic_set_bit32(spi->base_addr, SPI_DEVICE_CONTROL_REG_OFFSET,
+                                  SPI_DEVICE_CONTROL_ABORT);
+  while (!mmio_region_get_bit32(spi->base_addr, SPI_DEVICE_STATUS_REG_OFFSET,
+                                SPI_DEVICE_STATUS_ABORT_DONE)) {
+  }
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_reset(const dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  dif_spi_device_result_t err;
+
+  err = dif_spi_device_irq_clear_all(spi);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+  err = dif_spi_device_set_irq_levels(spi, 0x80, 0x00);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+
+  // Disable interrupts manually, since the relevant DIF only allows for single
+  // bit flips.
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_INTR_ENABLE_REG_OFFSET, 0);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_get(
+    const dif_spi_device_t *spi, dif_spi_device_irq_type_t type,
+    dif_spi_device_irq_state_t *state_out) {
+  if (spi == NULL || state_out == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t bit_index = 1;
+  switch (type) {
+    case kDifSpiDeviceIrqTypeRxFull:
+      bit_index = SPI_DEVICE_INTR_STATE_RXF;
+      break;
+    case kDifSpiDeviceIrqTypeRxAboveLevel:
+      bit_index = SPI_DEVICE_INTR_STATE_RXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeTxBelowLevel:
+      bit_index = SPI_DEVICE_INTR_STATE_TXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeRxError:
+      bit_index = SPI_DEVICE_INTR_STATE_RXERR;
+      break;
+    case kDifSpiDeviceIrqTypeRxOverflow:
+      bit_index = SPI_DEVICE_INTR_STATE_RXOVERFLOW;
+      break;
+    case kDifSpiDeviceIrqTypeTxUnderflow:
+      bit_index = SPI_DEVICE_INTR_STATE_TXUNDERFLOW;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
```

Comment:
Since this function covers all enum variants, error checking is not needed according to the DIF-specific style guide.